File: mobile/android/air/SubModules/WalletCore/src/main/java/org/mytonwallet/app_air/walletcore/helpers/FeeEstimationHelpers.kt
package org.mytonwallet.app_air.walletcore.helpers

import org.mytonwallet.app_air.walletcore.models.MBlockchain
import org.mytonwallet.app_air.walletcore.models.SwapType
import org.mytonwallet.app_air.walletcore.moshi.MApiSwapAsset
import org.mytonwallet.app_air.walletcore.stores.TokenStore
import java.math.BigDecimal

class FeeEstimationHelpers private constructor() {
    data class NetworkFeeData(
        val chain: MBlockchain?,
        val isNativeIn: Boolean,
        val fee: BigDecimal?
    )

    companion object {
        fun networkFeeData(
            sellToken: MApiSwapAsset?,
            sellTokenIsOnChain: Boolean,
            swapType: SwapType,
            networkFee: Double?
        ): NetworkFeeData? {
            sellToken ?: return null

            val tokenInChain = sellToken.chain.let { MBlockchain.valueOf(it ?: "") }
            val nativeUserTokenIn = if (sellTokenIsOnChain) {
                TokenStore.getToken(tokenInChain.nativeSlug)
            } else {
                null
            }
            val isNativeIn = sellToken.slug == nativeUserTokenIn?.slug
            val chainConfigIn = tokenInChain.gas

            val fee: BigDecimal? = run {
                var value: BigDecimal? = BigDecimal.ZERO
                when {
                    networkFee != null && networkFee > 0 -> {
                        value = networkFee.toBigDecimal()
                    }

                    swapType == SwapType.ON_CHAIN -> {
                        value = chainConfigIn?.maxSwap ?: BigDecimal.ZERO
                    }

                    swapType == SwapType.CROSS_CHAIN_FROM_TON -> {
                        value = if (isNativeIn) {
                            chainConfigIn?.maxTransfer
                        } else {
                            chainConfigIn?.maxTransferToken
                        }
                    }
                }
                value
            }

            return NetworkFeeData(chain = tokenInChain, isNativeIn = isNativeIn, fee = fee)
        }
    }
}

File: mobile/ios/Air/SubModules/WalletCore/Helpers/FeeEstimationHelpers.swift
//
//  FeeEstimationHelpers.swift
//  WalletCore
//
//  Created by Sina on 10/23/24.
//

import WalletContext

public class FeeEstimationHelpers {
    private init() {}
    
    public static func networkFeeBigInt(sellToken: ApiToken?, swapType: SwapType, networkFee: Double?) -> NetworkFeeData? {
        guard let sellToken else {
            return nil
        }
        let tokenInChain = ApiChain(rawValue: sellToken.chain)
        let nativeUserTokenIn = sellToken.isOnChain == true ? TokenStore.tokens[tokenInChain?.tokenSlug ?? ""] : nil
        let isNativeIn = sellToken.slug == nativeUserTokenIn?.slug
        let chainConfigIn = tokenInChain?.gas
        let fee = {
            var value: BigInt = 0
            if chainConfigIn == nil {
                return value
            }
            
            if (networkFee ?? 0 > 0) {
                value = doubleToBigInt(networkFee!, decimals: nativeUserTokenIn?.decimals ?? 9)
            } else if (swapType == SwapType.inChain) {
                value = chainConfigIn?.maxSwap ?? 0
            } else if (swapType == SwapType.crossChainFromTon) {
                value = (isNativeIn == true ? chainConfigIn?.maxTransfer : chainConfigIn?.maxTransferToken) ?? 0
            }
            
            return value;
        }()
        return NetworkFeeData(chain: tokenInChain, isNativeIn: isNativeIn, fee: fee)
    }
}
